packages/runtime-core/src/apiCreateApp.ts

```diff
@@ -373,15 +373,24 @@ export function createAppAPI<HostElement>(
         }
       },
 
-      provide(key, value) {
-        if (__DEV__ && (key as string | symbol) in context.provides) {
+      provide(key, newVal) {
+        const oldValue = context.provides[key as string | symbol]
+        const isValueChanged = hasChanged(oldValue, newVal)
+
+        if (
+          __DEV__ &&
+          (key as string | symbol) in context.provides &&
+          isValueChanged
+        ) {
           warn(
             `App already provides property with key "${String(key)}". ` +
               `It will be overwritten with the new value.`,
           )
         }
 
-        context.provides[key as string | symbol] = value
+        if (isValueChanged) {
```

Comment:
`if (isValueChanged) {` could be merged with previous `if`

Reply:
Are you suggesting to put it inside an `else` block? Since the values are already the same, there is no need to perform the assignment again.

Reply:
no
```js
if (isValueChanged) {
    if (...) warn()
  
    context.provides[key as string | symbol] = newVal
}
```

Reply:
got it.